src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs

```diff
@@ -0,0 +1,146 @@
+using Raven.Server.Documents.Queries.Results;
+using Raven.Server.Documents.Queries.Timings;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow.Json;
+using Sparrow.Logging;
+using Voron.Impl;
+
+namespace Raven.Server.Documents.Indexes.Persistence.Corax
+{
+    public class CoraxIndexReadOperation : IndexReadOperationBase
+    {
+        private readonly IndexSearcher _indexSearcher;
+        private readonly CoraxQueryEvaluator _coraxQueryEvaluator;
+        private long _entriesCount = 0;
+        private const int _bufferSize = 2048;
+
+        public CoraxIndexReadOperation(Index index, Logger logger, Transaction readTransaction) : base(index, logger)
+        {
+            _indexSearcher = new IndexSearcher(readTransaction);
+            _coraxQueryEvaluator = new CoraxQueryEvaluator(_indexSearcher);
+        }
+
+        public override void Dispose()
+        {
+            _indexSearcher?.Dispose();
+        }
+
+        public override long EntriesCount() => _entriesCount;
+
+        public override IEnumerable<QueryResult> Query(IndexQueryServerSide query, QueryTimingsScope queryTimings, FieldsToFetch fieldsToFetch,
+            Reference<int> totalResults, Reference<int> skippedResults,
+            IQueryResultRetriever retriever, DocumentsOperationContext documentsContext, Func<string, SpatialField> getSpatialField, CancellationToken token)
+        {
+            var pageSize = query.PageSize;
+            var isDistinctCount = pageSize == 0 && query.Metadata.IsDistinct;
+            if (isDistinctCount)
+                pageSize = int.MaxValue;
+            var position = query.Start;
+            long readCounter = 0;
+
+            IQueryMatch result = _coraxQueryEvaluator.Search(query.Metadata, fieldsToFetch);
+            if (result == null)
+                yield break;
+
+            totalResults.Value = Convert.ToInt32(result.Count);
+            var ids = ArrayPool<long>.Shared.Rent(_bufferSize);
+            try
+            {
+                int read = 0;
+
+                //escaping already returned docs.
+                // Q: Can Corax have "skip" method inside? It's inefficient to copy values only for skipping.
+                int docsToLoad = pageSize;
+                if (position != 0)
+                {
+                    int emptyRead = position / _bufferSize;
+                    do
+                    {
+                        read = result.Fill(ids);
+                        readCounter += read;
+                        emptyRead--;
+                    } while (emptyRead > 0);
+
+                    position %= _bufferSize; // move into <0;_bufferSize> set.
+                    //I know there is a cost of copying this but it's max _bufferSize and make code much simpler.
+                    ids[position..read].CopyTo(ids, 0);
+                    read -= position;
+                    skippedResults.Value = -1; // docsToLoad * _bufferSize + position;
```

Comment:
??

Reply:
The case for the studio to try to read more values (pagination). Waiting for answer here:  https://github.com/ravendb/ravendb/pull/12689#discussion_r701899233